subsys/usb/usb_device.c

```diff
@@ -503,18 +513,53 @@ static bool set_endpoint(const struct usb_ep_descriptor *ep_desc)
 		ep_cfg.ep_addr, ep_cfg.ep_type, ep_cfg.ep_mps);
 
 	if (usb_dc_ep_configure(&ep_cfg) < 0) {
-		LOG_WRN("Failed to configure endpoint %x", ep_cfg.ep_addr);
+		return false;
 	}
 
 	if (usb_dc_ep_enable(ep_cfg.ep_addr) < 0) {
-		LOG_WRN("Failed to enable endpoint %x", ep_cfg.ep_addr);
+		return false;
 	}
 
 	usb_dev.configured = true;
 
 	return true;
 }
 
+/*
+ * @brief Reset configuration and disable endpoint
+ *
+ * This function resets endpoint configuration to
+ * default values and disables it.
+ *
+ * @param [in]  ep_desc Endpoint descriptor byte array
+ *
+ * @return true if successfully deconfigured and disabled
+ */
+static bool reset_endpoint(const struct usb_ep_descriptor *ep_desc)
+{
+	struct usb_dc_ep_cfg_data ep_cfg = {0};
+
+	ep_cfg.ep_addr = ep_desc->bEndpointAddress;
+	ep_cfg.ep_type = ep_desc->bmAttributes;
+
+	if (ep_desc->bmAttributes > USB_DC_EP_INTERRUPT) {
+		return false;
+	}
+
+	LOG_DBG("Reset endpoint 0x%x type %u",
+		ep_cfg.ep_addr, ep_cfg.ep_type);
+
```

Comment:
The transfers for this endpoint have to be canceled, usb_cancel_transfer(ep_cfg.ep_addr);

Reply:
That's a good point.
As discussed offline:
Taking into consideration planned change for transfers (#21490) there will be a need to call usb_transfer() when enabling the OUT endpoints. I will address your comments and leave this PR for now (will be back once we are done with transfers).